src/Data/List/Ext/Properties.agda

```diff
@@ -29,6 +44,183 @@ _⊎-cong_ : ∀ {a b c d} {A : Set a} {B : Set b} {C : Set c} {D : Set d} {k} 
 h ⊎-cong h' = (h M.⊎-cong h')
   where open import Data.Sum.Function.Propositional as M
 
+-- TODO: stdlib?
+AllPairs⇒≡∨R∨Rᵒᵖ : ∀ {ℓ ℓ'} {A : Set ℓ} {R : A → A → Set ℓ'} {a b l}
+                 → AllPairs R l → a ∈ l → b ∈ l → a ≡ b ⊎ R a b ⊎ R b a
+AllPairs⇒≡∨R∨Rᵒᵖ [] = λ ()
+AllPairs⇒≡∨R∨Rᵒᵖ (_ ∷ _) (here refl) (here refl) = inj₁ refl
+AllPairs⇒≡∨R∨Rᵒᵖ (x ∷ _) (here refl) (there b∈l) = inj₂ (inj₁ (lookup x b∈l))
+AllPairs⇒≡∨R∨Rᵒᵖ (x ∷ _) (there a∈l) (here refl) = inj₂ (inj₂ (lookup x a∈l))
+AllPairs⇒≡∨R∨Rᵒᵖ (x ∷ h) (there a∈l) (there b∈l) = AllPairs⇒≡∨R∨Rᵒᵖ h a∈l b∈l
+
+
+module _ {a} {A : Set a} where
+
+  x≤x⊔y : ∀ x y → x ≤ x ⊔ y
+  x≤x⊔y zero y = z≤n
+  x≤x⊔y (suc x) zero = s≤s (≤-reflexive refl)
+  x≤x⊔y (suc x) (suc y) = s≤s (x≤x⊔y x y)
+
+  x≤y⊔x : ∀ x y → x ≤ y ⊔ x
+  x≤y⊔x zero y = z≤n
+  x≤y⊔x (suc x) zero = s≤s (x≤y⊔x x zero)
+  x≤y⊔x (suc x) (suc y) = s≤s (x≤y⊔x x y)
+
+  -- maximum length of the lists in the given list of lists
+  maxlen : List (List A) → ℕ
+  maxlen ls = foldr (λ l n → length l ⊔ n) 0 ls
+
+  maxlenˡ : List (List A) → ℕ
+  maxlenˡ ls = foldl (λ n l → n ⊔ length l) 0 ls
+
+  maxlen∷ : {ls : List (List A)} → ∀ l → maxlen (l ∷ ls) ≡ length l ⊔ maxlen ls
+  maxlen∷ {ls} l = refl
+
+  maxlen≤∷ : {ls : List (List A)} → ∀ l → maxlen ls ≤ maxlen (l ∷ ls)
+  maxlen≤∷ {[]} _ = z≤n
+  maxlen≤∷ {l' ∷ ls} l = subst (maxlen (l' ∷ ls) ≤_)
+                               (sym (maxlen∷{l' ∷ ls} l))
+                               (x≤y⊔x (maxlen (l' ∷ ls)) (length l))
+
+  ∈-maxlen-≤ : {ls : List (List A)} → ∀ l → l ∈ ls → length l ≤ maxlen ls
+  ∈-maxlen-≤ {l' ∷ ls} .l' (here refl) = subst (λ x → length l' ≤ x) (sym (maxlen∷{ls} l')) (x≤x⊔y (length l') (maxlen ls))
+  ∈-maxlen-≤ {l' ∷ ls} l (there l∈) = ≤-trans (∈-maxlen-≤{ls} l l∈) (maxlen≤∷{ls} l')
+
+-- ⊔-operator : MaxOperator ≤-totalPreorder
+-- ⊔-operator = record
+--   { x≤y⇒x⊔y≈y = m≤n⇒m⊔n≡n
+--   ; x≥y⇒x⊔y≈x = m≥n⇒m⊔n≡m
+--   }
+
+-- ≤-totalPreorder : TotalPreorder 0ℓ 0ℓ 0ℓ
+-- ≤-totalPreorder = record
+--   { isTotalPreorder = ≤-isTotalPreorder
+--   }
+
+-- ≤-isTotalPreorder : IsTotalPreorder _≡_ _≤_
+-- ≤-isTotalPreorder = record
+--   { isPreorder = ≤-isPreorder
+--   ; total      = ≤-total
+--   }
+
+-- ≤-isPreorder : IsPreorder _≡_ _≤_
+-- ≤-isPreorder = record
+--   { isEquivalence = isEquivalence
+--   ; reflexive     = ≤-reflexive
+--   ; trans         = ≤-trans
+--   }
+
+-- ≤-total : Total _≤_
+-- ≤-total zero    _       = inj₁ z≤n
+-- ≤-total _       zero    = inj₂ z≤n
+-- ≤-total (suc m) (suc n) = Sum.map s≤s s≤s (≤-total m n)
+
+-- Total : Rel A ℓ → Set _
+-- Total _∼_ = Connex _∼_ _∼_
+
+-- Connex : REL A B ℓ₁ → REL B A ℓ₂ → Set _
```

Comment:
I don't know why this is here

Reply:
oops... I just had it there for reference while I was working on the module. (removed)